starlette/middleware/limits.py

```
@@ -0,0 +1,66 @@
+"""Middleware that limits the body size of incoming requests."""
+from starlette.datastructures import Headers
+from starlette.responses import PlainTextResponse
+from starlette.types import ASGIApp, Message, Receive, Scope, Send
+
+DEFAULT_MAX_BODY_SIZE = 2_621_440  # 2.5MB
+
+
+class ContentTooLarge(Exception):
+    def __init__(self, max_body_size: int) -> None:
+        self.max_body_size = max_body_size
+
+
+class LimitRequestMiddleware:
+    def __init__(
+        self, app: ASGIApp, max_body_size: int = DEFAULT_MAX_BODY_SIZE
+    ) -> None:
+        self.app = app
+        self.max_body_size = max_body_size
+
+    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
+        if scope["type"] != "http":  # pragma: no cover
+            return await self.app(scope, receive, send)
+
+        headers = Headers(scope=scope)
+        content_length = headers.get("content-length")
+        if content_length is not None:
+            if int(content_length) > self.max_body_size:
+                return await _content_too_large_app(scope)(scope, receive, send)
+
+            # NOTE: The server makes sure that the content-length header sent by the
+            #   client is the same as the length of the body.
+            #   Ref.: https://github.com/django/asgiref/issues/422
+            return await self.app(scope, receive, send)
```

I still need to do more thinking/research here. I don't think it makes sense for the middleware to do this validation.

On the other hand... Is just 1 conditional...

---

Doesn’t hurt to do!

---

Done.

---

To be clear, I think the advantage of this is that if the client says the body is 1GB and our limit is 1MB we can return an error without reading 1MB of the body into memory. If the client says the body is 1KB we should still verify that's the case. So it's a fail-fast path.